### ISSUE TITLE
Attributes not removed when set to null on updates

### ISSUE THREAD

**Issue by steadicat** [facebook/react#1448]
The behavior of attributes with a value of `null` or `undefined` is inconsistent. On first render they are not present on the element at all. On subsequent renders, the attribute sticks around with an empty value.

For example, `<a href={null} />` renders what you expect (`<a />`), but if the value of `href` changes to `null` after the first render, you get `<a href />`. In the case of `href` (and likely others) an empty attribute has unintended side effects.

See this fiddle: http://jsfiddle.net/pGG2A/1/

**Comment by ezequiel** [comment]
@syranide @sebmarkbage 
>\<progress /\> is different from <progress value="" />, but that's probably the only exception (I know of at >least). CSS-rules is an obvious one I guess, although extremely rarely used it seems, but there are >some legit uses.

Apparently the `dir` attribute may be part of the exception as well. See http://jsbin.com/pasicazexe/3/edit. 

Things to note: The `html` element's `dir` attribute has been set `rtl`, and the `input` element's `dir` attribute has been set to `""` (no value). The problem is the above html **behaves differently in Chrome and Firefox.** I suppose this is expected, as the only [valid values for `dir` are `ltr`, `rtl`, and `auto`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/dir). I assume the behavior of the value `""` (no value) is undefined, or perhaps the browser ignores it (Chrome doesn't seem to in this case).

For now, my component has this in place:

```js
componentDidUpdate: function() {
    if (!this.props.dir) {
        this.getDOMNode().removeAttribute('dir');
    }
}
```

**Comment by syranide** [comment]
@ezequiel Ah good to know.

**Comment by abrgr** [comment]
I'm also getting bit by this, specifically for hrefs on anchor tags.  Here is a minimal test case of the very unexpected behavior: https://jsfiddle.net/vsa3Ln40/1/

I'm happy to make the changes if the maintainers agree that the current behavior is unexpected enough to warrant a change.

**Comment by tgecho** [comment]
`href` on anchor tags is an issue for me as well. When the href is set to `null` or `undefined` the link remains active, and clicking on it has the effect of removing anything after `#` in the url. My temporary workaround is to add a key to the jsx output of my custom link component.

```
<a href={href} key={href ? 'y' : 'n'} />
```

**Comment by yaycmyk** [comment]
I opened a bug against the IE team for this: https://connect.microsoft.com/IE/feedbackdetail/view/1525237/removeattribute-slowness-compared-to-other-browsers

**Comment by syranide** [comment]
#1510 has landed in 0.15 and this is practically solved (except for a special-case #1431).

**Comment by robozb** [comment]
Why React removes included attributes from any html tag autonomously? It can cause css selection problem :(
